File: leetcode/206-reverse-linked-list.ts
// The problem can be found on https://leetcode.com/problems/two-sum

// * Definition for singly-linked list.
class ListNode {
   val: number
   next: ListNode | null
   constructor(val?: number, next?: ListNode | null) {
      this.val = val === undefined ? 0 : val
      this.next = next === undefined ? null : next
   }
}

/**
 * Runtime: 0ms, Memory: 52.74MB
 *
 * Time: O(n), Space: O(n).
 * First attempt:
 */
function reverseList(head: ListNode | null): ListNode | null {
   if (head === null || !head.next) {
      return head
   }

   let prev: ListNode | null = null

   while (head !== null) {
      prev = new ListNode(head.val, prev)
      head = head.next
   }

   return prev
}

/**
 * Time: O(n), Space: O(1)
 * Note: Not my solution
 */
function reverseList2(head: ListNode | null): ListNode | null {
   if (head === null || !head.next) {
      return head
   }

   let prev: ListNode | null = null
   while (head) {
      let next = head.next
      head.next = prev
      prev = head
      head = next
   }
   return prev
}

File: leetcode/linked-lists/206-reverse-linked-list.go
package main

// Definition for singly-linked list.
type ListNode struct {
   Val int
   Next *ListNode
}

// Runtime: 0ms, Memory: 4.68MB
func reverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
  	}

  	var prev *ListNode = nil

  	for head != nil {
		prev = &ListNode{
			 Val:  head.Val,
			 Next: prev,
		}
		head = head.Next
  	}

  	return prev
}

// Runtime: 0, Memory: 4.49MB
func reverseList2(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
	  	return head
  	}
  	var prev *ListNode
  	for head != nil {
	  	temp := head.Next
	  	head.Next = prev
	  	prev = head
	  	head = temp
  	}
  	return prev
}